modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/AbstractGeoIpDownloader.java

```diff
@@ -0,0 +1,166 @@
+        // the high default poll interval of 3d), and it simplifies the logic considerably.
+        requestRunOnDemand();
+
+        synchronized (this) {
+            scheduledPeriodicRun = threadPool.schedule(this::runPeriodic, pollIntervalSupplier.get(), threadPool.generic());
+        }
+    }
+
+    /**
+     * This method requests that the downloader runs on the latest cluster state, which likely contains a change in the GeoIP metadata.
+     * This method does nothing if this task is cancelled or completed.
+     */
+    public void requestRunOnDemand() {
+        if (isCancelled() || isCompleted()) {
+            logger.debug("Not requesting downloader to run on demand because task is cancelled or completed");
+            return;
+        }
+        logger.trace("Requesting downloader run on demand");
+        // If queuedRuns was greater than 0, then either a run is in progress and it will fire off another run when it finishes,
+        // or a run is scheduled to run as soon as possible and it will include the latest cluster state.
+        // If it was 0, we set it to 1 to indicate that a run is scheduled to run as soon as possible and schedule it now.
+        if (queuedRuns.getAndIncrement() == 0) {
+            logger.trace("Scheduling downloader run on demand");
+            threadPool.generic().submit(this::runOnDemand);
+        }
+    }
+
+    /**
+     * Runs the downloader on the latest cluster state. {@link #queuedRuns} protects against multiple concurrent runs and ensures that
+     * if a run is requested while this method is running, then another run will be scheduled to run as soon as this method finishes.
+     */
+    private void runOnDemand() {
+        if (isCancelled() || isCompleted()) {
+            logger.debug("Not running downloader on demand because task is cancelled or completed");
+            return;
+        }
+        // Capture the current queue size, so that if another run is requested while we're running, we'll know at the end of this method
+        // whether we need to run again.
+        final int currentQueueSize = queuedRuns.get();
+        logger.trace("Running downloader on demand");
+        try {
+            runDownloader();
+            logger.trace("Downloader completed successfully");
+        } finally {
+            // If any exception was thrown during runDownloader, we still want to check queuedRuns.
+            // Subtract this "batch" of runs from queuedRuns.
+            // If queuedRuns is still > 0, then a run was requested while we were running, so we need to run again.
+            if (queuedRuns.addAndGet(-currentQueueSize) > 0) {
+                logger.debug("Downloader on demand requested again while running, scheduling another run");
+                threadPool.generic().submit(this::runOnDemand);
+            }
+        }
+    }
+
+    /**
+     * Download, update, and clean up GeoIP databases as required by the GeoIP processors in the cluster.
```

Comment:
nit: might be nice to know without scanning code

```suggestion
     * Download, update, and clean up GeoIP databases as required by the GeoIP processors in the cluster.
     * Guaranteed to not be called concurrently.
```